.evergreen/happy-eyeballs-server.py

```diff
@@ -0,0 +1,145 @@
+parser.add_argument('--stop', action='store_true', help='stop a currently-running server')
+parser.add_argument('--wait', action='store_true', help='wait for a server to be ready')
+args = parser.parse_args()
+
+PREFIX='happy eyeballs server'
+
+async def control_server():
+    shutdown = asyncio.Event()
+    srv = await asyncio.start_server(lambda reader, writer: on_control_connected(reader, writer, shutdown), 'localhost', args.control)
+    print(f'{PREFIX}: listening for control connections on {args.control}', file=sys.stderr)
+    async with srv:
+        await shutdown.wait()
+    print(f'{PREFIX}: all done', file=sys.stderr)
+
+async def on_control_connected(reader: asyncio.StreamReader, writer: asyncio.StreamWriter, shutdown: asyncio.Event):
+    # Read the control request byte
+    data = await reader.readexactly(1)
+    if data == b'\x04':
+        print(f'{PREFIX}: ========================', file=sys.stderr)
+        print(f'{PREFIX}: request for delayed IPv4', file=sys.stderr)
+        slow = 'IPv4'
+    elif data == b'\x06':
+        print(f'{PREFIX}: ========================', file=sys.stderr)
+        print(f'{PREFIX}: request for delayed IPv6', file=sys.stderr)
+        slow = 'IPv6'
+    elif data == b'\xF0':
+        writer.write(b'\x01')
+        await writer.drain()
+        writer.close()
+        await writer.wait_closed()
+        return
+    elif data == b'\xFF':
+        print(f'{PREFIX}: shutting down', file=sys.stderr)
+        writer.close()
+        await writer.wait_closed()
+        shutdown.set()
+        return
+    else:
+        print(f'Unexpected control byte: {data}', file=sys.stderr)
+        exit(1)
+    
+    # Bind the test ports but do not yet start accepting connections
+    connected = asyncio.Event()
+    on_ipv4_connected = lambda reader, writer: on_test_connected('IPv4', writer, b'\x04', connected, slow)
+    on_ipv6_connected = lambda reader, writer: on_test_connected('IPv6', writer, b'\x06', connected, slow)
+    # port 0: pick random unused port
+    srv4 = await asyncio.start_server(on_ipv4_connected, 'localhost', 0, family=socket.AF_INET, start_serving=False)
+    srv6 = await asyncio.start_server(on_ipv6_connected, 'localhost', 0, family=socket.AF_INET6, start_serving=False)
+    ipv4_port = srv4.sockets[0].getsockname()[1]
+    ipv6_port = srv6.sockets[0].getsockname()[1]
+    print(f'{PREFIX}: [slow {slow}] open for IPv4 on {ipv4_port}', file=sys.stderr)
+    print(f'{PREFIX}: [slow {slow}] open for IPv6 on {ipv6_port}', file=sys.stderr)
+
+    # Reply to control request with success byte and test server ports
+    writer.write(b'\x01')
+    writer.write(ipv4_port.to_bytes(2, 'big'))
+    writer.write(ipv6_port.to_bytes(2, 'big'))
+    await writer.drain()
+    writer.close()
+    await writer.wait_closed()
+
+    # Start test servers listening in parallel
+    # Hold a reference to the tasks so they aren't GC'd
```

Comment:
Other fun fact: there's a better API for this idiom (`asyncio.TaskGroup`) but it's also not in all evergreen python installations.